lib/ddtrace/auto_instrument.rb

```diff
@@ -0,0 +1,41 @@
+require 'ddtrace'
+
+module Datadog
+  # AutoInstrumentation enables all integrations
+  module AutoInstrument
+    module_function
+
+    def patch_all
+      integrations = []
+      Datadog.registry.each do |integration|
+        # some instrumentations are automatically enabled when the `rails` instrumentation is enabled,
+        # patching them on their own automatically outside of the rails integration context would
+        # cause undesirable service naming, so we exclude them based their auto_instrument? setting.
+        # we also don't want to mix rspec/cucumber integration in as rspec is env we run tests in.
+        next unless integration.klass.auto_instrument?
+        integrations << integration.name
+      end
+
+      Datadog.configure do |c|
+        c.reduce_log_verbosity
+        # This will activate auto-instrumentation for Rails
+        integrations.each do |integration_name|
+          c.use integration_name
+        end
+      end
+    end
+
+    if Datadog::Contrib::Rails::Utils.railtie_supported?
+      # Railtie to include AutoInstrumentation in rails loading
+      class Railtie < Rails::Railtie
+        # we want to load before config initializers so that any user supplied config
+        # in config/initializers/datadog.rb will take precedence
+        initializer 'datadog.start_tracer', before: :load_config_initializers do
+          AutoInstrument.patch_all
+        end
+      end
```

Comment:
_Following up on @delner's suggestion, about separating core from contrib._

We have 3 parts to this file which I think we can break apart:

1. This Rails-specific snippet. Can be extracted to the Rails sub-directory under contrib.
1. The `patch_all` implementation. This parts knows how to patch all instrumentations, but does not trigger itself. I don't know 100% where this would go, but `Datadog::Contrib::Extensions` looks very similar to where it could land. `Datadog::Contrib::Extensions` might be too overloaded for it, but I think it give us a good idea to where "integration core" features live today.
    To avoid coupling today `Datadog::Contrib::Extensions` injects itself into the global settings object: `Datadog::Configuration::Settings.send(:include, Configuration::Settings)`. This is a valid technique to prevent coupling.
1. A "product" part, where we activate it all. This is the code that is ultimately called when we require `ddtrace/auto_instrument`.

To make 1 and 2 work, one possible suggestion is to introduce another functionality to integrations: similarly to how they support `#patch` today, they could support something like `auto_patch`, and we could call that on all integrations that support it. This way new integration can choose to provide custom auto instrumentation or not.